### PR TITLE
[UXIT-1538] Event Page · Integrate Luma events via API

### PR DESCRIPTION
**Pull request by mirhamasala**
## 📝 Description
This PR replaces the static Luma iframe embed with a dynamic card-based events display that fetches data directly from the Luma API. This provides a more integrated and customizable event display experience while maintaining fallback to the iframe when no events are available.

- **Type:** New feature

## 🛠️ Key Changes
- Created new Luma API integration service to fetch events data
- Added Zod schema validation for Luma API responses
- Implemented new card-based UI for displaying events
- Added fallback to iframe when no events are available
- Created utility functions for date formatting and metadata mapping

## 📌 To-Do Before Merging
- [ ] Verify `NEXT_LUMA_FILECOIN_API_KEY` is configured in environment
- [ ] Test error handling scenarios
- [ ] Verify fallback to iframe works correctly
- [ ] Review caching strategy (currently set to 1 hour revalidation)
- [ ] Refactor `getEventMetaData` from `src/app/_utils/getMetaData.ts` (in a different PR?)

## 🧪 How to Test
- **Setup:** 
  - Ensure `NEXT_LUMA_FILECOIN_API_KEY` environment variable is set
  - Run the development server

- **Steps to Test:**
  1. Navigate to an events page with Luma integration
  2. Verify events are displayed in card format with correct metadata
  3. Test with API key removed to verify error handling
  4. Test with empty events response to verify iframe fallback
  
- **Expected Results:** 
  - Events should display in a responsive card grid
  - Each card should show event name, description, date, and location
  - Clicking "View Event Details" should link to the Luma event page
  - When no events are available, should fallback to iframe or "No events" message

## 🔖 Resources
- [Luma API Documentation](https://lu.ma/api)
- [Zod Documentation](https://zod.dev/)
- [Next.js Data Fetching Documentation](https://nextjs.org/docs/basic-features/data-fetching)

CleanShot 2024-11-01 at 11 46 18@2x

### CODE REVIEW

File: src/app/events/[slug]/services/lumaService.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    const rawData = await res.json()
+
+    const validatedData = LumaResponseSchema.parse(rawData)
+
+    return validatedData.entries
+  } catch (error) {
+    if (error instanceof z.ZodError) {
+      console.error('Invalid API response format:', error.errors)
```
Comment (CharlyMartin) [inline src/app/events/[slug]/services/lumaService.ts:33]: Should we use `logZodError` here?

File: src/app/events/[slug]/components/EventsSection.tsx [hunk 2]
```diff
@@ -1,18 +1,47 @@
...
+              metaData={getMetaData(mapLumaEventToMetadata(event))}
+              textIsClamped={true}
+              cta={{
+                text: 'View Event Details',
+                href: event.url,
+              }}
+            />
+          ))}
+        </CardGrid>
+      ) : embedLink ? (
```
Comment (CharlyMartin) [inline src/app/events/[slug]/components/EventsSection.tsx:35]: Not a big fan of double ternary operators, I think we can do better than that :)

The pattern we usually follow is we use `try catch` in the component fetching the data:
if success -> render the list
if success but data is empty -> show empty state
if error - offer a fallback (the iframe in our case)

File: src/app/events/[slug]/services/lumaService.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import { z } from 'zod'
+
+import { type LumaEventEntry, LumaResponseSchema } from '../schemas/LumaSchema'
+
+export async function getLumaEvents(): Promise<Array<LumaEventEntry>> {
+  const apiKey = process.env.NEXT_LUMA_FILECOIN_API_KEY
+  if (!apiKey) throw new Error('Luma API key is not configured')
+
+  try {
```
Comment (CharlyMartin) [inline src/app/events/[slug]/services/lumaService.ts:9]: The `try catch` might be unnecessary here, depending on how we want to handle the error state.

File: src/app/events/[slug]/services/lumaService.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+          accept: 'application/json',
+          authorization: `Bearer ${apiKey}`,
+        },
+        next: { revalidate: 3600 },
+      },
+    )
+
+    if (!res.ok) {
+      throw new Error(`Failed to fetch events: ${res.status} ${res.statusText}`)
+    }
```
Comment (CharlyMartin) [inline src/app/events/[slug]/services/lumaService.ts:24]: Might be unnecessary. `await res.json()` and/or `LumaResponseSchema.parse(rawData)` will throw is something goes wrong.

File: src/app/events/[slug]/components/EventsSection.tsx [hunk 5]
```diff
@@ -1,18 +1,47 @@
+import { z } from 'zod'
+
```
Comment (mirhamasala) [inline src/app/events/[slug]/components/EventsSection.tsx:2]:
```suggestion
```

File: src/app/events/[slug]/services/lumaService.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
+import { z } from 'zod'
+
+import { type LumaEventEntry, LumaResponseSchema } from '../schemas/LumaSchema'
+
+export async function getLumaEvents(): Promise<Array<LumaEventEntry>> {
```
Comment (CharlyMartin) [inline src/app/events/[slug]/services/lumaService.ts:5]: TypeScript already knows that is returns `Promise<Array<LumaEventEntry>>`.

File: src/app/events/[slug]/services/lumaService.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  const apiKey = process.env.NEXT_LUMA_FILECOIN_API_KEY
```
Comment (CharlyMartin) [inline src/app/events/[slug]/services/lumaService.ts:6]: Why is the ENV name prefixed with NEXT?

**Review by CharlyMartin — CHANGES_REQUESTED** [review]
Mirha, it works but I think we can re-organise the code to follow our usual server-side data fetching pattern.

File: src/app/events/[slug]/services/lumaService.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+  try {
+    const res = await fetch(
+      'https://api.lu.ma/public/v1/calendar/list-events',
+      {
+        method: 'GET',
+        headers: {
+          accept: 'application/json',
+          authorization: `Bearer ${apiKey}`,
+        },
+        next: { revalidate: 3600 },
```
Comment (CharlyMartin) [inline src/app/events/[slug]/services/lumaService.ts:18]: Nice!

**Comment by mirhamasala** [comment]
## Closing this PR: LUMA API Integration Limitations

We're closing this PR due to fundamental limitations with the LUMA API that prevent us from achieving our intended functionality.

### Key Issues Discovered:

1. **Limited Event Visibility**
   - The `/calendar/list-events` endpoint only returns events *managed* by our calendar
   - Events that are *listed* but not *managed* by us are not accessible via the API
   - This means we'd miss the majority of our community events (currently only ~2 events would show up)

2. **Lack of Tag Filtering**
   - The API doesn't provide functionality to filter events by tags
   - This would be crucial for separating events as we scale

### Alternative Solution Considered:

**Manual CMS Management**
   - Could manually add events to our CMS
   - While technically feasible, this creates additional maintenance overhead and content team prefers to keep pulling events from LUMA

### Decision:

We'll continue using the LUMA embed solution as it:
- Shows all events (both managed and listed) correctly
- Maintains real-time synchronization with LUMA
- Requires no additional maintenance overhead
- Provides consistent UX across all event calendars

### Future Considerations:

We can revisit this integration if:
- LUMA expands their API to include listed events
- They add tag filtering capabilities
- Our requirements change regarding event display and management

**Pull request closed without merging**